FastRTC Release Post (#2698)
huggingface/blog: fastrtc.md

---
title: "FastRTC: The Real-Time Communication Library for Python" 
thumbnail: /blog/assets/fastrtc/fastrtc_logo.png
authors:
- user: freddyaboulton
---

# FastRTC: The Real-Time Communication Library for Python

> In the last six months, the AI audio space has exploded with model releases (for both open and closed source models) and investor and developer interest. To name a few milestones:

Reviewer: ```suggestion
In the last few months, many new real-time speech models have been released and entire companies have been founded (around both open and closed source models). To name a few milestones:
```

- OpenAI and Google released their live multimodal APIs for ChatGPT and Gemini. OpenAI even went so far as to release a 1-800-ChatGPT phone number!

> - Kyutai released Moshi, a fully open-source audio-to-audio LLM. Alibaba released Qwen2-Audio and Fixie.ai released Ultravox - two open-source LLMs that natively understand audio.

Reviewer: Link to models on Hub

> - EleveLabs raised $180m in their Series C.

Reviewer: ```suggestion
- [ElevenLabs](https://elevenlabs.io/) raised $180m in their Series C.
```

> Despite the explosion in the model and funding side, it's still difficult to build real-time AI applications, especially in Python.

Reviewer: ```suggestion
Despite the explosion in the model and funding side, it's still difficult to build real-time AI applications that stream audio or video, especially in Python.
```

> - ML engineers may not have experience with the technologies needed to build real-time applications.

Reviewer: ```suggestion
- ML engineers may not have experience with the technologies needed to build real-time applications, such as WebRTC.
```

- Even code assistant tools like Cursor and Copilot struggle to write python code that supports real-time audio/video applications. I know from experience!

> That's why we're excited to announce `FastRTC`, the real-time communication library for Python. The library is designed to make it super easy to build real-time audio and video AI applications entirely in Python! Let's dive in.

Reviewer: I would show an example code snippet here to illustrate how simple FastRTC is to use, before the Core Features.

Reviewer: Then it'll be more natural to talk about the `Stream` class

Author: I removed mentions of `Stream` and merged the core features with the introduction. I don't want them to go after code because they motivate what the code will show about fastrtc.

> - Even code assistant tools like Cursor and Copilot struggle to write python code that supports real-time audio/video applications. I know from experience!

Reviewer: ```suggestion
- Even code assistant tools like Cursor and Copilot struggle to write Python code that supports real-time audio/video applications. I know from experience!
```

- 💻 Automatic UI - Built-in webRTC-enabled [Gradio](https://www.gradio.app/) UI for testing (or deploying to prod!).
- 📞 Call via Phone - Use fastphone() to get a FREE phone number to call into your audio stream (HF Token required. Increased limits for PRO accounts).
- ⚡️ WebRTC and Websocket support
- 💪 Customizable - You can mount the stream to any FastAPI app so you can serve a custom UI or deploy beyond Gradio
- 🧰 Lots of utils for text-to-speech, speech-to-text, stop word detection to get you started

## Getting Started

Let's start into the "hello word" of real-time audio: echoing back what the user says. In `FastRTC`, this is as simple as:

```python
from fastrtc import Stream, ReplyOnPause
import numpy as np

def echo(audio: tuple[int, np.ndarray]) -> tuple[int, np.ndarray]:
    yield audio

stream = Stream(ReplyOnPause(echo), modality="audio", mode="send-receive")
stream.ui.launch()
```

Let's break it down:
- The `ReplyOnPause` will handle the voice detection and turn taking for you. You just have to worry about the logic for responding to the user. Any generator that returns a tuple of audio, (represented as `(sample_rate, audio_data)`) will work.

> - The `Stream` class will build a production-ready Gradio UI for you to quickly test out your stream (or deploy to prod!).

Reviewer: The "deploy to prod" part is not clear to me. I think I would phrase it something like, "once you are ready prototyping, you can deploy your Stream as a production-ready FastAPI app in a single line of code"

Here it is in action:

<video src="https://github.com/user-attachments/assets/fcf2d30e-3e98-47c9-8dc3-23340784c441" controls /></video>

## Leveling-Up: LLM Voice Chat

The next level is to use an LLM to respond to the user. `FastRTC` comes with built-in speech-to-text and text-to-speech capabilities so working with LLMs is really easy. Let's change our `echo` function accordingly:

```python
import os

from fastrtc import (ReplyOnPause, Stream, get_stt_model, get_tts_model)
from openai import OpenAI

sambanova_client = OpenAI(
    api_key=os.getenv("SAMBANOVA_API_KEY"), base_url="https://api.sambanova.ai/v1"
)
stt_model = get_stt_model()
tts_model = get_tts_model()

def echo(audio):
    prompt = stt_model.stt(audio)
    response = sambanova_client.chat.completions.create(
        model="Meta-Llama-3.2-3B-Instruct",
        messages=[{"role": "user", "content": prompt}],
        max_tokens=200,
    )
    prompt = response.choices[0].message.content
    for audio_chunk in tts_model.stream_tts_sync(prompt):
        yield audio_chunk

stream = Stream(ReplyOnPause(echo), modality="audio", mode="send-receive")
stream.ui.launch()
```

> We're using the SambaNova API since it's fast. But you can use any LLM/text-to-speech/speech-to-text API. Bring the tools you love - `FastRTC` just handles the real-time communication layer.

Reviewer: ```
stt_model = get_stt_model()
tts_model = get_tts_model()
```
are doing some heavy lifting. I would mention what they are doing and that you can replace them with your own STT/TTS models are skip them altogether if you use a voice-to-voice model/api
